### PR TITLE
YLP-15 Change CV calculation

### PR DESCRIPTION
**Pull request by franck-fourel**
The CV is calculated as an average of the daily CV. 
We calculate the daily CV if we have at least 3 values in the day.

### CODE REVIEW

File: src/utils/data.js [hunk 1]
```diff
@@ -261,22 +277,22 @@ export class DataUtil {
...
-      total,
-    };
+    } = this.getStandardDevData(true);
 
     if (insufficientData) {
-      coefficientOfVariationData.insufficientData = true;
+      return {
+        insufficientData,
+        total,
+        coefficientOfVariation: NaN,
```
Comment (yvon-dblg) [inline src/utils/data.js]:
```suggestion
        coefficientOfVariation: Number.NaN,
```

File: src/utils/data.js [hunk 2]
```diff
@@ -425,25 +441,40 @@ export class DataUtil {
...
       return {
         averageGlucose,
         insufficientData: true,
-        standardDeviation: NaN,
         total,
+        standardDeviation: NaN,
       };
     }
-
+
```
Comment (yvon-dblg) [inline src/utils/data.js]:
```suggestion

```

File: src/utils/data.js [hunk 3]
```diff
@@ -425,25 +441,40 @@ export class DataUtil {
...
+    const { bgData, averageGlucose, bgDataByDate, total } = this.getGlucoseDataByDate(true);
 
-    if (bgData.length < 3) {
+    if (total < 3) {
       return {
         averageGlucose,
         insufficientData: true,
-        standardDeviation: NaN,
         total,
+        standardDeviation: NaN,
```
Comment (yvon-dblg) [inline src/utils/data.js]:
```suggestion
        standardDeviation: Number.NaN,
```

File: package-lock.json [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
-  "version": "0.14.0",
+  "version": "0.14.1",
   "lockfileVersion": 1,
   "requires": true,
   "dependencies": {
     "@babel/cli": {
       "version": "7.10.3",
       "resolved": "https://nexus.ci.diabeloop.eu/repository/npm-diabeloop/@babel/cli/-/cli-7.10.3.tgz",
-      "integrity": "sha512-lWB3yH5/fWY8pi2Kj5/fA+17guJ9feSBw5DNjTju3/nRi9sXnl1JPh7aKQOSvdNbiDbkzzoGYtsr46M8dGmXDQ==",
+      "integrity": "sha1-Tqg72ZfSpBx40HJjraPsRm+zdks=",
```
Comment (yvon-dblg) [inline package-lock.json]: Too much unwanted change in this file.
Please update your npm version.

File: src/utils/data.js [hunk 5]
```diff
@@ -171,6 +171,22 @@ export class DataUtil {
...
+  /**
+   * Returns the average glucose data 
+   * with the average glucose data by date
+   */
+  getGlucoseDataByDate = () => {
+    const data = this.getAverageGlucoseData(true);
+
+    const bgDataByDate = _.groupBy(data.bgData, function(d) { 
+      return d.localDate; 
+    });
```
Comment (yvon-dblg) [inline src/utils/data.js]:
```suggestion
    const bgDataByDate = _.groupBy(data.bgData, d => d.localDate);
```

File: src/utils/data.js [hunk 6]
```diff
@@ -425,25 +441,40 @@ export class DataUtil {
     };
   };
 
-  getStandardDevData = () => {
-    const { averageGlucose, bgData, total } = this.getAverageGlucoseData(true);
+  getStandardDevData = ( cvByDate = false) => {
+    const { bgData, averageGlucose, bgDataByDate, total } = this.getGlucoseDataByDate(true);
```
Comment (yvon-dblg) [inline src/utils/data.js]:
```suggestion
    const { bgData, averageGlucose, bgDataByDate, total } = this.getGlucoseDataByDate();
```

File: src/utils/data.js [hunk 7]
```diff
@@ -425,25 +441,40 @@ export class DataUtil {
...
+    
     const squaredDiffs = _.map(bgData, d => (d.value - averageGlucose) ** 2);
     const standardDeviation = Math.sqrt(_.sum(squaredDiffs) / (bgData.length - 1));
 
+    const coefficientOfVariationByDate = {};
+    if (cvByDate) {
+      _.forEach(bgDataByDate, function(value, key) {
+        // ignore days having less than 3 glycemia values
+        if (value.length >= 3 ) {
+          const avgGlucose =  _.meanBy(value, 'value');
```
Comment (yvon-dblg) [inline src/utils/data.js]:
```suggestion
          const avgGlucose = _.meanBy(value, 'value');
```

File: src/utils/data.js [hunk 8]
```diff
@@ -425,25 +441,40 @@ export class DataUtil {
     };
   };
 
-  getStandardDevData = () => {
-    const { averageGlucose, bgData, total } = this.getAverageGlucoseData(true);
+  getStandardDevData = ( cvByDate = false) => {
```
Comment (yvon-dblg) [inline src/utils/data.js]:
```suggestion
  getStandardDevData = (cvByDate = false) => {
```

File: src/utils/data.js [hunk 9]
```diff
@@ -425,25 +441,40 @@ export class DataUtil {
...
       };
     }
-
+    
     const squaredDiffs = _.map(bgData, d => (d.value - averageGlucose) ** 2);
     const standardDeviation = Math.sqrt(_.sum(squaredDiffs) / (bgData.length - 1));
 
+    const coefficientOfVariationByDate = {};
+    if (cvByDate) {
+      _.forEach(bgDataByDate, function(value, key) {
```
Comment (yvon-dblg) [inline src/utils/data.js]:
```suggestion
      _.forEach(bgDataByDate, (value, key) => {
```

File: src/utils/data.js [hunk 10]
```diff
@@ -425,25 +441,40 @@ export class DataUtil {
...
+          const standardDeviation = Math.sqrt(_.sum(squaredDiffs) / (value.length - 1));
+          coefficientOfVariationByDate[key] = standardDeviation / avgGlucose * 100;
+        }
+      })
+    }
+
     return {
-      averageGlucose,
-      standardDeviation,
+      averageGlucose,
```
Comment (yvon-dblg) [inline src/utils/data.js]:
```suggestion
      averageGlucose,
```

File: src/utils/data.js [hunk 11]
```diff
@@ -171,6 +171,22 @@ export class DataUtil {
     return data;
   };
 
+  /**
+   * Returns the average glucose data
```
Comment (yvon-dblg) [inline src/utils/data.js]:
```suggestion
   * Returns the average glucose data
```

File: src/utils/data.js [hunk 12]
```diff
@@ -425,25 +441,40 @@ export class DataUtil {
...
+          coefficientOfVariationByDate[key] = standardDeviation / avgGlucose * 100;
+        }
+      })
+    }
+
     return {
-      averageGlucose,
-      standardDeviation,
+      averageGlucose,      
+      insufficientData: false,
```
Comment (yvon-dblg) [inline src/utils/data.js:471]: Should it be true if we have less than 3 days of data?
See your comment:
> `// ignore days having less than 3 glycemia values`

File: test/utils/data.test.js [hunk 13]
```diff
@@ -715,20 +778,39 @@ describe('DataUtil', () => {
...
     });
 
-    it('should return `NaN` when less than 3 datums available', () => {
+    it('should return true for insufficientData when less than 3 datums available', () => {
       dataUtil = new DataUtil(smbgData.slice(0, 2), opts({ chartPrefs: { bgSource: 'smbg' } }));
       expect(dataUtil.getCoefficientOfVariationData()).to.eql({
-        coefficientOfVariation: NaN,
         insufficientData: true,
         total: 2,
+        coefficientOfVariation: NaN,
```
Comment (yvon-dblg) [inline test/utils/data.test.js]:
```suggestion
        coefficientOfVariation: Number.NaN,
```

File: src/utils/data.js [hunk 14]
```diff
@@ -425,25 +441,40 @@ export class DataUtil {
...
-
+    
     const squaredDiffs = _.map(bgData, d => (d.value - averageGlucose) ** 2);
     const standardDeviation = Math.sqrt(_.sum(squaredDiffs) / (bgData.length - 1));
 
+    const coefficientOfVariationByDate = {};
+    if (cvByDate) {
+      _.forEach(bgDataByDate, function(value, key) {
+        // ignore days having less than 3 glycemia values
+        if (value.length >= 3 ) {
```
Comment (yvon-dblg) [inline src/utils/data.js]:
```suggestion
        if (value.length >= 3) {
```

File: src/utils/data.js (hunk 9 above)
Comment (yvon-dblg) [reply src/utils/data.js]: value & key are generic (default) variable names
Having more explicit names, and/or comment of their type/content and/or [typescript tags](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html) could be helpful.

File: src/utils/data.js [hunk 15]
```diff
@@ -425,25 +441,40 @@ export class DataUtil {
...
+          const avgGlucose =  _.meanBy(value, 'value');
+          const squaredDiffs = _.map(value, d => (d.value - avgGlucose) ** 2);
+          const standardDeviation = Math.sqrt(_.sum(squaredDiffs) / (value.length - 1));
+          coefficientOfVariationByDate[key] = standardDeviation / avgGlucose * 100;
```
Comment (yvon-dblg) [inline src/utils/data.js:464]: Comments are needed on this formula.
The math formula is needed for me, at least in the Jira
Something like:
image

**Review by yvon-dblg — CHANGES_REQUESTED** [review]
- Few minors changes
- Need more documentation to understand the formula

File: src/utils/data.js (hunk 15 above)
Comment (franck-fourel) [reply src/utils/data.js:464]: JIRA updated

**Review by yvon-dblg — APPROVED** [review]
LGTM

**Pull request merged**